app/src/main/java/com/android/sample/feature/schedule/repository/schedule/FirestoreScheduleRepository.kt

```
@@ -0,0 +1,176 @@
+
+  private val collectionName = "schedule"
+  private val usersCollection = "users"
+
+  init {
+    startListening()
+  }
+
+  private fun startListening() {
+    val uid = auth.currentUser?.uid ?: return
+
+    db.collection(usersCollection).document(uid).collection(collectionName).addSnapshotListener {
+        snapshot,
+        error ->
+      if (error != null || snapshot == null) {
+        // You might log the error here
+        return@addSnapshotListener
+      }
+
+      val events =
+          snapshot.documents
+              .mapNotNull { it.toScheduleEvent() }
+              .sortedWith(compareBy<ScheduleEvent> { it.date }.thenBy { it.time ?: LocalTime.MIN })
+
+      _events.value = events
+    }
+  }
+
+  private fun userScheduleCollection(uid: String) =
+      db.collection(usersCollection).document(uid).collection(collectionName)
+
+  private fun requireUid(): String {
+    return auth.currentUser?.uid
+        ?: throw IllegalStateException("No authenticated user for ScheduleRepository")
+  }
+
+  override suspend fun save(event: ScheduleEvent) {
+    // Treat save as an upsert
+    val uid = requireUid()
+    val collection = userScheduleCollection(uid)
+
+    val id = if (event.id.isBlank()) collection.document().id else event.id
+    val eventWithId = event.copy(id = id)
+
+    collection.document(id).set(eventWithId.toFirestoreMap(), SetOptions.merge()).await()
+    // No need to update _events manually; snapshot listener will fire.
+  }
+
+  override suspend fun update(event: ScheduleEvent) {
+    val uid = requireUid()
+    val id =
+        event.id.ifBlank {
+          throw IllegalArgumentException("update() requires an event with a non-blank id")
+        }
+
+    userScheduleCollection(uid).document(id).set(event.toFirestoreMap(), SetOptions.merge()).await()
+  }
+
+  override suspend fun delete(eventId: String) {
+    val uid = requireUid()
+    if (eventId.isBlank()) return
```

if (eventId.isBlank()) return. Silent failure is bad UX. Return Result.Failure with error message.